[WIP]Added cases for sparse arrays in tensorproduct

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#16941

#### Brief description of what is fixed or changed
Sparse array are cast into list while performing tensorproduct. Brief overview in #16941
**Before:**
```python
>>> a = MutableSparseNDimArray.zeros(10000, 20000)
>>> res = tensorproduct(a, 3)
Too slow
```
Thanks to the lazy-evaluation, it would not cause a MemoryError immediately. But as the evaluation continues, it is supposed to go beyond the limit of memory. So it should be fixed.
**Now:**

```python
>>> a = MutableSparseNDimArray({1:1}, (3, ))
>>> b = MutableSparseNDimArray({2:1}, (3, ))

>>> tensorproduct(a, b) == ImmutableSparseNDimArray({5: 1}, (3, 3))
True
```
#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added specific cases for sparse array in tensorproduct and tests

<!-- END RELEASE NOTES -->

Commit:
Added some cases and tests

- Added case for tensorproduct of a scalar and a sparse array
- Added tests for sparse array in test of tensorproduct

kangzhiq commented on sympy/tensor/array/tests/test_arrayop.py:
```diff
@@ -43,6 +43,11 @@ def test_tensorproduct():
 
+    # tests for large scale sparse array
+    a = MutableSparseNDimArray.zeros(10000, 20000)
+    assert tensorproduct(a, x) == MutableSparseNDimArray.zeros(10000, 20000)
```
Test will not pass because equality test for large scale sparse array is not fixed yet.(Referring to #16937)

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -55,6 +57,14 @@ def tensorproduct(*args):
+    if isinstance(b, SparseNDimArray) and not isinstance(a, NDimArray):
+        return type(b)({k: a*v for (k, v) in b._sparse_array}, b.shape)
+
+    # TODO: product for two sparse arrays
```
don't we only need the product of two sparse arrays?

kangzhiq replied:
You are right. I will work on the case of two sparse arrays. Thank you for reviewing!

Commit:
Added cases for sp_arr for tensorproduct

- Removed unnecessary codes
- Added case for tensorproduct between two sp_arr
- Added tests

Commit:
Update with master and fix conflict

Upabjojr commented:
Can you add a for loop testing both dense and sparse classes?

e.g. instead o
```python
X = Array(...)
assert tensorproduct( ... ) == ...
```
write:
```python
for ArrayType in [ImmutableDenseArray, ImmutableSparseArray]:
    X = ArrayType( ... )
    assert tensorproduct( ... ) == ...
```

In this way we are sure all tests are repeated for both dense and sparse arrays.

Commit:
Modified tests in test_arrayop

- Added loop to make sure test passed for sprase and dense arrays
- Ameliorated tests in test_derive_by_array
